### PR TITLE
[feature]: Add component-based develop Backbone Network for EasyRec

### PR DESCRIPTION
**Pull request by yangxudong**

### CODE REVIEW

File: easy_rec/python/utils/dag.py [hunk 1]
```diff
@@ -0,0 +1,205 @@
...
+      if not isinstance(dep_nodes, list):
+        raise TypeError('dict values must be lists')
+      for dep_node in dep_nodes:
+        self.add_edge(ind_node, dep_node)
+
+  def reset_graph(self):
+    """Restore the graph to an empty state."""
+    self.graph = OrderedDict()
+
+  def ind_nodes(self, graph=None):
```
Comment (poson) [inline easy_rec/python/utils/dag.py:142]: ind 太不直观了

File: easy_rec/python/utils/dag.py [hunk 2]
```diff
@@ -0,0 +1,205 @@
...
+    else:
+      raise ValueError('graph is not acyclic')
+
+  def size(self):
+    return len(self.graph)
+
+
+if __name__ == '__main__':
+  dag = DAG()
+  dag.add_node('a')
```
Comment (poson) [inline easy_rec/python/utils/dag.py:196]: 这里应该挪到test再添加几个测试

File: easy_rec/python/layers/keras/bst.py [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+
+    seq_embed_size = seq_input.shape.as_list()[-1]
+    if seq_embed_size != self.config.hidden_size:
+      seq_input = tf.layers.dense(
+          seq_input,
+          self.config.hidden_size,
+          activation=tf.nn.relu,
+          kernel_regularizer=self.l2_reg)
+
+    # seq_len: [batch_size, 1], the true length of each sequence
```
Comment (lgqfhwy) [inline easy_rec/python/layers/keras/bst.py:83]: 这些注释掉的东西不用的话可以删除

File: easy_rec/python/layers/keras/bst.py [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+      # target_feature: [batch_size, 1, embed_size]
+      target_feature = tf.expand_dims(target_feature, 1)
+      # seq_input: [batch_size, seq_len+1, embed_size]
+      seq_input = tf.concat([target_feature, seq_input], axis=1)
+
+    return self.encode(seq_input, max_position)
+
+  # def contrastive_loss(self, seq_input, seq_len, max_position):
+  #   aug_seq1, aug_seq2, aug_len1, aug_len2 = input_aug_data(seq_input, seq_len)
+  #   seq_output1 = self.encode(aug_seq1, max_position)
```
Comment (lgqfhwy) [inline easy_rec/python/layers/keras/bst.py:119]: 这些注释的不用的也可以删除

File: easy_rec/python/layers/keras/din.py [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+import logging
+
+import tensorflow as tf
+from tensorflow.python.keras.layers import Layer
+
+from easy_rec.python.layers import dnn
+from easy_rec.python.utils.shape_utils import get_shape_list
+
+
+class DIN(Layer):
```
Comment (lgqfhwy) [inline easy_rec/python/layers/keras/din.py:12]: 这个 din 支持 召回模型 和 负采样 一起用吗，没看到有对应的测试

File: .git_bin_url [hunk 6]
```diff
@@ -1,5 +1,4 @@
 {"leaf_path": "data/test", "sig": "656d73b4e78d0d71e98120050bc51387", "remote_path": "data/git_oss_sample_data/data_test_656d73b4e78d0d71e98120050bc51387"}
-{"leaf_path": "data/test/client", "sig": "d2e000187cebd884ee10e3cf804717fc", "remote_path": "data/git_oss_sample_data/data_test_client_d2e000187cebd884ee10e3cf804717fc"}
```
Comment (chengmengli06) [inline .git_bin_url:2]: should not be deleted

File: docs/source/index.rst [hunk 7]
```diff
@@ -37,7 +44,6 @@ Welcome to easy_rec's documentation!
    :maxdepth: 2
    :caption: TRAIN & EVAL & EXPORT
 
-   loss
```
Comment (chengmengli06) [inline docs/source/index.rst:40]: should be kept

File: easy_rec/python/layers/input_layer.py [hunk 8]
```diff
@@ -92,8 +94,12 @@ def __call__(self, features, group_name, is_combine=True, is_dict=False):
     feature_name_to_output_tensors = {}
     negative_sampler = self._feature_groups[group_name]._config.negative_sampler
     if is_combine:
-      concat_features, group_features = self.single_call_input_layer(
-          features, group_name, feature_name_to_output_tensors)
+      place_on_cpu = os.getenv('place_embedding_on_cpu')
```
Comment (chengmengli06) [inline easy_rec/python/layers/input_layer.py:97]: if necessary, put it in a separate PR

File: easy_rec/python/layers/input_layer.py [hunk 9]
```diff
@@ -123,12 +129,24 @@ def __call__(self, features, group_name, is_combine=True, is_dict=False):
       group_columns, group_seq_columns = feature_group.select_columns(
           self._fc_parser)
 
-      assert len(group_columns) == 0, \
-          'there are none sequence columns: %s' % str(group_columns)
+      embedding_reg_lst = []
```
Comment (chengmengli06) [inline easy_rec/python/layers/input_layer.py:132]: explain this piece of code, why is it necessary?

File: easy_rec/python/model/dbmtl.py [hunk 10]
```diff
@@ -37,24 +37,27 @@ def __init__(self,
                                          features,
                                          self._model_config.bottom_uniter,
                                          self._input_layer)
-    else:
-      self._features, _ = self._input_layer(self._feature_dict, 'all')
+    elif not self.has_backbone:
+      self._features, self._feature_list = self._input_layer(
+          self._feature_dict, 'all')
```
Comment (chengmengli06) [inline easy_rec/python/model/dbmtl.py:42]: no else?

File: easy_rec/python/layers/input_layer.py [hunk 11]
```diff
@@ -140,7 +158,7 @@ def __call__(self, features, group_name, is_combine=True, is_dict=False):
           embedding_reg_lst.append(tmp_embedding)
       regularizers.apply_regularization(
           self._embedding_regularizer, weights_list=embedding_reg_lst)
-      return seq_features
+      return seq_features, output_features, group_features
```
Comment (chengmengli06) [inline easy_rec/python/layers/input_layer.py:161]: this interface is already too complicated, consider refactor it into a new function

File: easy_rec/python/model/easy_rec_model.py [hunk 12]
```diff
@@ -47,7 +48,7 @@ def __init__(self,
     self._emb_reg = regularizers.l2_regularizer(self.embedding_regularization)
     self._l2_reg = regularizers.l2_regularizer(self.l2_regularization)
     # only used by model with wide feature groups, e.g. WideAndDeep
-    self._wide_output_dim = -1
+    self._wide_output_dim = self.get_wide_output_dim()
```
Comment (chengmengli06) [inline easy_rec/python/model/easy_rec_model.py:51]: it is only used by wide_and_deep, may be put in WideAndDeep model only, should not be placed in base model

**Review by chengmengli06 — APPROVED** [review]
